boot_shutdown_interface/src/boot_shutdown_interface.cpp

```diff
@@ -0,0 +1,149 @@
+// Copyright 2022 TIER IV, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "boot_shutdown_interface/boot_shutdown_interface.hpp"
+
+namespace boot_shutdown_interface
+{
+
+BootShutdownInterface::BootShutdownInterface(const rclcpp::NodeOptions & options)
+: Node("boot_shutdown", options),
+  ecu_name_(declare_parameter("ecu_name", std::string("ecu"))),
+  startup_timeout_(declare_parameter("startup_timeout", 180)),
+  prepare_shutdown_time_(declare_parameter("prepare_shutdown_time", 1)),
+  execute_shutdown_time_(declare_parameter("execute_shutdown_time", 13))
+{
+  using namespace std::literals::chrono_literals;
+  using std::placeholders::_1;
+  using std::placeholders::_2;
+
+  ecu_state_.name = ecu_name_;
+  ecu_state_.state = EcuState::STARTUP;
+  startup_time_ = get_clock()->now();
+
+  srv_prepare_shutdown_ = create_service<PrepareShutdown>(
+    "~/service/prepare_shutdown", std::bind(
+      &BootShutdownInterface::onPrepareShutdown, this, _1,
+      _2),
+    rmw_qos_profile_default);
+
+  srv_execute_shutdown_ = create_service<ExecuteShutdown>(
+    "~/service/execute_shutdown", std::bind(
+      &BootShutdownInterface::onExecuteShutdown, this, _1,
+      _2),
+    rmw_qos_profile_default);
+
+  pub_ecu_state_ =
+    create_publisher<EcuState>("~/output/ecu_state", rclcpp::QoS{1});
+
+  auto qos = rclcpp::QoS{1};
+  if (declare_parameter(
+      "topic_config.transient_local",
+      static_cast<rclcpp::ParameterValue>(false)).get<bool>())
+  {
+    qos.transient_local();
+  }
+  if (declare_parameter(
+      "topic_config.best_effort",
+      static_cast<rclcpp::ParameterValue>(false)).get<bool>())
+  {
+    qos.best_effort();
+  }
+
+  sub_topic_ = create_generic_subscription(
+    declare_parameter("topic_config.name", rclcpp::PARAMETER_STRING).get<std::string>(),
+    declare_parameter("topic_config.type", rclcpp::PARAMETER_STRING).get<std::string>(),
+    qos, std::bind(&BootShutdownInterface::onTopic, this, std::placeholders::_1));
+
+  timer_ = rclcpp::create_timer(
+    this, get_clock(), 1s, std::bind(&BootShutdownInterface::onTimer, this));
+}
+
+void BootShutdownInterface::onPrepareShutdown(
+  PrepareShutdown::Request::SharedPtr, PrepareShutdown::Response::SharedPtr response)
+{
+  RCLCPP_INFO(get_logger(), "shutdown preparing...");
+
+  // Clear pagecache, dentries, and inodes.
+  preparation_child_ = boost::process::child(
```

Comment:
@ito-san sorry for sudden commend,
for this boost, will it be better to use popen?
As some of the ecu do not have boost installed

Reply:
@tkhmy 
Wecome! Thank you for your comment.
Yes, I would avoid dependencies on boost libraries and `popen` is good solution.
However, I would like to standardized the system as below near the future in all ECUs like this.
![image](https://user-images.githubusercontent.com/57388357/197463045-85277964-936c-4a3d-be76-a200be9fd750.png)
This design uses UNIX domain socket to communicate with ROS node and daemon process, and boost serialization library is used for serialized communication. What do you think about it?

Reply:
Sorry, I'm not really understand, I thought the serialized call will be handle by ros2 service?
If boost serialization library is used in this design, I think the other ecu also will need to have boost lib?

Reply:
@tkhmy 
The serialized communication between ros node and daemon process is handle by socket communication like this, not ros2 service.
https://github.com/tier4/boot_shutdown_tools/blob/7b77fdcaa30a1ac2a94bfa4ac388348dceab03ce/boot_shutdown_interface/src/boot_shutdown_guest.cpp#L184-L197
> If boost serialization library is used in this design, I think the other ecu also will need to have boost lib?

Yes, you're correct.

Reply:
Yes, I just found the code in #2 PR, so I will install boost lib for the ecu